particle-iot/device-os
# [rtl872x] Fix pmic shared interrupt. Fix usb detection, minor

### Problem

VIN on TrackerM hardware on `5.3.0`/`develop` is broken again. 

The power source also always shows as `VIN` when on USB power. 
This is due to `HAL_PLATFORM_POWER_WORKAROUND_USB_HOST_VIN_SOURCE` define. `HAL_USB_Get_State()` is not implemented / always returns `HAL_USB_STATE_NONE`. This forces the `VIN` source even if the PMIC reports USB source due to the following:
```
    switch (powerSourceFromStatus(status)) {
      case POWER_SOURCE_USB_HOST: {
#if HAL_PLATFORM_POWER_WORKAROUND_USB_HOST_VIN_SOURCE
        if (vin || usb_state < HAL_USB_STATE_POWERED) { // <------ This is always true, so SRC is always set to `POWER_SOURCE_VIN`
#else
        if (vin) {
#endif // HAL_PLATFORM_POWER_WORKAROUND_USB_HOST_VIN_SOURCE
          src = POWER_SOURCE_VIN;
          applyVinConfig();
        }
    ...
```

### Solution

Fix the shared interrupt config version to match latest in HAL.
Remove `HAL_PLATFORM_POWER_WORKAROUND_USB_HOST_VIN_SOURCE` from RTL platform 

### Steps to Test

Run test app below and switch from VIN / USB / Battery, the power sources / power change events should be logged on `Serial1`

### Example App

```
Serial1LogHandler logHandler(115200, LOG_LEVEL_WARN, {
    { "app", LOG_LEVEL_TRACE },
    { "sys.power", LOG_LEVEL_TRACE },
    { "system", LOG_LEVEL_TRACE },
});

const char * map_power_source (int power_source_enum) {
    // POWER_SOURCE_UNKNOWN = 0,
    // POWER_SOURCE_VIN = 1,
    // POWER_SOURCE_USB_HOST = 2,
    // POWER_SOURCE_USB_ADAPTER = 3,
    // POWER_SOURCE_USB_OTG = 4,
    // POWER_SOURCE_BATTERY = 5
    const char * power_sources[] = {
        "UNKNOWN",
        "VIN",
        "USB_HOST",
        "USB_ADAPTER",
        "USB_OTG",
        "BATTERY",
    };

    if (power_source_enum <=5) {
        return power_sources[power_source_enum];
    } else {
        return power_sources[0];
    }
}

const char * map_battery_state (int battery_state_enum) {
    // BATTERY_STATE_UNKNOWN = 0,
    // BATTERY_STATE_NOT_CHARGING = 1,
    // BATTERY_STATE_CHARGING = 2,
    // BATTERY_STATE_CHARGED = 3,
    // BATTERY_STATE_DISCHARGING = 4,
    // BATTERY_STATE_FAULT = 5,
    // BATTERY_STATE_DISCONNECTED = 6
    const char * battery_states[] = {
        "UNKNOWN",
        "NOT_CHARGING",
        "CHARGING",
        "CHARGED",
        "DISCHARGING",
        "FAULT",
        "DISCONNECTED"
    };

    if(battery_state_enum <=6) {
        return battery_states[battery_state_enum];
    } else {
        return battery_states[0];
    }
}

void power_source_changed_handler(system_event_t event, int powerSource) {
    Log.warn("Event %u: POWER SOURCE CHANGED TO --> %s", (uint32_t)event, map_power_source(powerSource));
}

static int previousPowerSource, previousBatteryState = 0;

/* This function is called once at start up ----------------------------------*/
void setup()
{
#if PLATFORM_ID == PLATFORM_TRACKERM
    // Enable Serial1 UART Voltage translator chip
    pinMode(A0, OUTPUT);
    digitalWrite(A0, HIGH);
    pinMode(A5, OUTPUT);
    digitalWrite(A5, HIGH);

#endif

    System.on(SystemEvents::power_source, power_source_changed_handler);
}

/* This function loops forever --------------------------------------------*/
void loop()
{
    //This will run in a loop
    delay(1000);

    FuelGauge fuel;

    int powerSource = System.powerSource();
    int batteryState = System.batteryState();
    float batterySoc = System.batteryCharge();

    float batteryVoltage = fuel.getVCell();
    bool alertStatus = fuel.getAlert();

    if ((powerSource != previousPowerSource) || (batteryState != previousBatteryState)) {
        Log.info("PowerSource=%s BatteryState=%s BatVoltage=%.2f BatSOC=%.1f alertStatus=%d",
            map_power_source(powerSource),
            map_battery_state(batteryState),
            batteryVoltage,
            batterySoc,
            alertStatus);
    }
    previousPowerSource = powerSource;
    previousBatteryState = batteryState;    

    if (Serial1.available()) {
        char c = Serial1.read();
        if (c == 'r') {
            // Force update event
            Log.info("Get/Set power config");
            const auto config = System.getPowerConfiguration();        
            System.setPowerConfiguration(config);
        } else if (c == 's') {
            Log.info("State: PowerSource=%s BatteryState=%s BatVoltage=%.2f BatSOC=%.1f alertStatus=%d",
                map_power_source(powerSource),
                map_battery_state(batteryState),
                batteryVoltage,
                batterySoc,
                alertStatus);
        }
    }
}

```

### References

Links to the Community, Docs, Other Issues, etc..

---

### Completeness
